### DISCUSSION TITLE
gatsby-transformer-asciidoc - Allow Macro Extensions

### DISCUSSION THREAD

**Discussion by bitPogo** [gatsbyjs/gatsby#27510] [Ideas / Feature Requests]
## Summary

As a user of gatsby-transformer-asciidoc I like to be able to use my awesome custom asciidoc [extenstions](gatsby-transformer-asciidoc).

### Basic example

```javascript
// taken from https://asciidoctor-docs.netlify.app/asciidoctor.js/extend/extensions/inline-macro-processor/
// **somewhere**/myAwesomeExtension.js
module.exports = function (registry) {
  registry.inlineMacro('emoticon', function () {
    var self = this
    self.process(function (parent, target) {
      var text
      if (target === 'grin') {
        text = ':D'
      } else if (target === 'wink') {
        text = ';)'
      } else {
        text = ':)'
      }
      return self.createInline(parent, 'quoted', text, { 'type': 'strong' }).convert()
    })
  })
}
````
```javascript
 // In your gatsby-config.js, make sure to import or declare TemplateConverter
MyAwsomeExtention = require('**somewhere**/myAwesomeExtension.js')
...
plugins: [
  {
    resolve: `gatsby-transformer-asciidoc`,
    options: {
      extensionFactory: MyAwesomeExtension,
    },
  },
]
```
### Motivation

I am actually in need of more elaborate customisation, which exceeds the capability of a CustomConverter unless I do it in a hacky way, which I do not want to. A custom macro extension allows me to achieve that without a bigger effort. It also opens up gatsby-transformer-asciidoc for a more convenient customisation in general.

### Offer

Since it relatively easy to do that, I can also provide a PR by myself, if it is wanted.

**Comment by iainbryson** [comment]
Hi @bitPogo I might need something like this. have you already done the work? Does it work with Opal-generated extensions too?

**Reply by bitPogo** [reply]
> Hi @bitPogo I might need something like this. have you already done the work?

The PR is open (see #27514).

> Does it work with Opal-generated extensions too?

That's a really good question...if ascidoctorjs supports that...it should work, since the plugin will just delegate the extensions to asciidoctorjs.
